Modeling HVAC Operational Faults – Coil Supply Air Temperature Sensor Offset

rongpengzhang opened the pull request:
## Pull request overview

The fault modeling features aim to model the HVAC operational faults to improve the accuracy of building simulations and to support the building retrofitting. This branch deals with the 4th fault in the following fault list to be implemented in Phase III. 
1.  Chiller Supply Water Temperature Sensor Offset
2.  Condenser Supply Water Temperature Sensor Offset
3.  Cooling Tower Scaling
4.  Coil Supply Air Temperature Sensor Offset 

Note that this branch is created from Commit 920b6cc94a93d41f2931f5a99a9d831a8565c1a6 in Branch #5766. 
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/120694345_FaultModel_CondensorSWT_TowerFouling' into 120694415_FaultModel_CoilSATSensor

rongpengzhang committed:
V_c1. IDD design for the CoilSATSensorOffset

rongpengzhang committed:
V_c2. Define in hh file the new class for this fault

rongpengzhang committed:
V_c3. Fault Struct initialization for CoilSATSensorOffset

rongpengzhang committed:
V_c4_a. Apply the fault model for Coil:Heating:Steam

rongpengzhang committed:
V_c4_b. Apply the fault model for Coil:Heating types

rongpengzhang committed:
V_c4_c. Apply the fault model for three water coil types

rongpengzhang committed:
Example file

rongpengzhang committed:
Documentation

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

mjwitte commented on src/EnergyPlus/WaterCoils.hh:
```diff
@@ -273,6 +273,10 @@ namespace WaterCoils {
 		bool DesiccantRegenerationCoil; // true if it is a regeneration air heating coil defined in Desiccant Dehumidifier system
 		int DesiccantDehumNum; // index to desiccant dehumidifier object
+		// variables for coil fault
+		bool FaultyCoilSATFlag; // True if the coil has SAT sensor fault
```
@rongpengzhang I don't think these new variables are used, because the faults are applied in the water coil controller.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -80385,6 +80389,121 @@ FaultModel:EnthalpySensorOffset:ReturnAir,
+       \key Coil:Heating:Steam
+       \key Coil:Heating:Water
+       \key Coil:Cooling:Water
+       \key Coil:Cooling:Water:Detailedgeometry
```
@rongpengzhang What about CoilSystem:Cooling:DX and CoilSystem:Heating:DX?  These are both temperature-based coil controls.  Also, AirloopHVAC:UnitarySystem can serve a similar function (acting as just a coil controller with a setpoint).

rongpengzhang replied:
Done.

mjwitte commented on tst/EnergyPlus/unit/FaultsManager.unit.cc:
```diff
@@ -115,25 +116,32 @@ TEST_F( EnergyPlusFixture, FaultsManager_FaultFoulingAirFilters_CheckFaultyAirFi
+	TestRestult = FaultsFouledAirFilters( FanNum ).CheckFaultyAirFilterFanCurve();
 	EXPECT_FALSE( TestRestult );
 
 	// Clean up
```
@rongpengzhang Is there a new unit test for the coil supply temp fault?

mjwitte reviewed (changes requested):
@rongpengzhang @Myoldmopar In general, the methodology for the coil supply air temperature sensor offset appears to be sound.  But it is difficult to review because the other new fault models are also in this branch.  So, this has to be merged last, after the other ones drop.  I will review again after the others are in so I can see just the changes for coil supply air temperature offset.

mjwitte commented:
@rongpengzhang The three operational fault test files are failing with input errors.  Probably due to late transition changes that dropped into v8.6.

rongpengzhang committed:
Address comments

Remove fault information from WaterCoils.hh

rongpengzhang committed:
Unit test

rongpengzhang committed:
Example File update

rongpengzhang committed:
Apply the fault model to CoilSystem:Cooling:DX

rongpengzhang committed:
Apply the fault model to CoilSystem:Heating:DX

rongpengzhang committed:
Apply the fault model to AirLoopHVAC:UnitarySystem

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

rongpengzhang committed:
Slight modification on fault application to unitary system

The fault model is only applicable to the unitary system controlled on
leaving air temperature

rongpengzhang commented:
@mjwitte @Myoldmopar  I have finished the code updates to address Mike's new comments, including: (1) apply the fault model to CoilSystem:Cooling:DX, CoilSystem:Heating:DX, and AirloopHVAC:UnitarySystem that can be controlled based on leaving air temperature, (2) develop a unit test for the feature, and (3) update the example files from V8.5 to V8.6. Can you please review the updates? Thanks!

nrel-bot-2 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 20 days since this pull request was last updated.

rongpengzhang committed:
Solve merge conflicts

# Conflicts:
#	doc/engineering-reference/src/operational-faults.tex
#	doc/input-output-reference/src/overview/group-operational-faults.tex
#	src/EnergyPlus/FaultsManager.cc
#	src/EnergyPlus/FaultsManager.hh
#	testfiles/CMakeLists.txt
#	tst/EnergyPlus/unit/FaultsManager.unit.cc

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

# Conflicts:
#	doc/engineering-reference/src/operational-faults.tex
#	doc/input-output-reference/src/overview/group-operational-faults.tex
#	src/EnergyPlus/FaultsManager.cc
#	src/EnergyPlus/FaultsManager.hh
#	testfiles/CMakeLists.txt
#	tst/EnergyPlus/unit/FaultsManager.unit.cc

mjwitte commented:
@rongpengzhang @Myoldmopar I've reviewed and approved the requested changes to add the additional coil types, but I have not tested any of these yet.

rongpengzhang committed:
Solve merge conflicts

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

rongpengzhang committed:
Address merge conflicts 1

# Conflicts:
#	doc/engineering-reference/src/operational-faults.tex
#	doc/input-output-reference/src/overview/group-operational-faults.tex
#	testfiles/CMakeLists.txt

rongpengzhang committed:
Address merge conflicts 2

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694415_FaultModel_CoilSATSensor

# Conflicts:
#	doc/engineering-reference/src/operational-faults.tex
#	doc/input-output-reference/src/overview/group-operational-faults.tex
#	testfiles/CMakeLists.txt

rongpengzhang committed:
Merge branch '120694415_FaultModel_CoilSATSensor' of https://github.com/NREL/EnergyPlus into 120694415_FaultModel_CoilSATSensor

rongpengzhang commented:
@Myoldmopar  I have addressed the new merge conflicts caused by the merge of #5766. Let's see the CI test results. If there are no errors, this branch is clean to review. This is the last one of FY16 fault models.
